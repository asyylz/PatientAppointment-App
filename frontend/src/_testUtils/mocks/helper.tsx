import { within } from '@testing-library/react';
import { render, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { Store } from '@reduxjs/toolkit';
import React from 'react';

export const toContainRole = async (
  container: HTMLElement,
  role: string,
  quantity = 1
): Promise<{ pass: boolean; message: () => string }> => {
  const elements = await within(container).findAllByRole(role);


  if (elements.length === quantity) {
    return {
      pass: true,
      message: () => '', // return an empty string when the test passes
    };
  }

  return {
    pass: false,
    message: () =>
      `Expected to find ${quantity} ${role} elements. Found ${elements.length} instead.`,
  };
};

export const renderComponent = (
  component: React.ReactElement,
  store: Store,
  options?: RenderOptions
) => {
  return render(
    <Provider store={store}>
      <MemoryRouter>{component}</MemoryRouter>
    </Provider>,
    options
  );
};

// export const toContainRole = (
//   container: HTMLElement,
//   role: string,
//   quantity = 1
// ): { pass: boolean; message: () => string } => {
//   const elements = within(container).getAllByRole(role);
// console.log(elements.length)
//   if (elements.length === quantity) {
//     return {
//       pass: true,
//       message: () => '', // return an empty string when the test passes
//     };
//   }

//   return {
//     pass: false,
//     message: () =>
//       `Expected to find ${quantity} ${role} elements. Found ${elements.length} instead.`,
//   };
// };
