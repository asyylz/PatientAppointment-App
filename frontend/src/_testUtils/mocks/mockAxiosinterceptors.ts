import { axiosInterceptorsWithToken } from './../../services/axiosInterceptors'
/* ---------------- Mock axiosInterceptor --------------- */
jest.mock('./../../hooks/axiosInterceptors', () => ({
  axiosInterceptorsWithToken: {
    delete: jest.fn(),
    post: jest.fn(),
    patch: jest.fn(),
    get: jest.fn(),
    interceptors: {
      request: { use: jest.fn() },
      response: { use: jest.fn() },
    },
  },
}));

// Mock delete
const deleteMock = axiosInterceptorsWithToken.delete as jest.Mock;
deleteMock.mockResolvedValue({
  data: { success: true }, // Simulate a successful delete response
});

// Mock update
const updateMock = axiosInterceptorsWithToken.patch as jest.Mock;
updateMock.mockResolvedValue({
  status: 'success',
  data: {
    appointment: {
      _id: '66bb4c6a7f78041ac61a875c',
      doctorId: '665f0f2959c971659af920e5',
      patientId: '6673662fbd42a966b75dec92',
      appointmentDateAndTime: '2024-08-16T16:00:00.000Z',
      reason: 'ccccc',
      diagnose: null,
      status: null,
      referral: false,
      __v: 0,
    },
  },
});

// Mock get
const getMock = axiosInterceptorsWithToken.get as jest.Mock;
getMock.mockImplementationOnce(() => new Promise(() => {}));
getMock.mockResolvedValue({
  data: {
    status: 'succeeded',
    error: null,
    data: {
      total: 1,
      upcomingAppointments: 1,
      appointmentsForPatient: [
        {
          _id: '66b0a0c31d2124f7d8aaa65b',
          doctorId: {
            _id: '665f0f2959c971659af920ca',
            firstName: 'Bowen',
            lastName: 'Chan',
            departmentId: '6660fbd12e3c00fe18cfceef',
          },
          patientId: '6673662fbd42a966b75dec92',
          appointmentDateAndTime: '2024-08-16T11:00:00.000Z',
          reason: 'test3',
          diagnose: null,
          status: null,
          referral: false,
        },
      ],
    },
  },
});
