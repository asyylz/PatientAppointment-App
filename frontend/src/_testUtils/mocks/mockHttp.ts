import useHttp from "../../hooks/useHttp/useHttp";

const deleteAppointmentMock = jest.fn();
// we have response conditon in Dashboard component to be able to empty the states
deleteAppointmentMock.mockResolvedValue({
  status: 204, // Simulate a successful deletion response with status 204
});

const updateAppointmentMock = jest.fn();
// we have response conditon in Dashboard component to be able to empty the states
updateAppointmentMock.mockResolvedValue({
  status: 'success',
});
const createAppointmentMock = jest.fn();
createAppointmentMock.mockResolvedValue({
  status: 'success',
});

const submitContactFormMock = jest.fn();
submitContactFormMock.mockResolvedValue({
  status: 'success',
});
const getDoctorWithAvailabilitiesMock = jest.fn();
getDoctorWithAvailabilitiesMock.mockResolvedValue({
  status: 'success',
});

(useHttp as jest.Mock).mockReturnValue({
  deleteAppointment: deleteAppointmentMock,
  updateAppointment: updateAppointmentMock,
  createAppointment: createAppointmentMock,
  getDoctorWithAvailabilities: getDoctorWithAvailabilitiesMock,
  submitContactForm: submitContactFormMock,
});

export {
  deleteAppointmentMock,
  updateAppointmentMock,
  createAppointmentMock,
  submitContactFormMock,
  getDoctorWithAvailabilitiesMock,
};
