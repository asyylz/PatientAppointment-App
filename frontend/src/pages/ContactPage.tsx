import React, { useState } from 'react';
import CustomInput from '../components/UI/CustomInput';
import classes from './ContactPage.module.css';

interface contactForm {
  name: string;
  email: string;
  subject: string;
  message: string;
}
const ContactPage: React.FC = () => {
  const [contactData, setContactData] = useState<contactForm>({
    name: '',
    email: '',
    subject: '',
    message: '',
  });

  const handleCancelClick = () => {
    setContactData({
      name: '',
      email: '',
      subject: '',
      message: '',
    });
  };
  const handleInputChange = (
    e: React.ChangeEvent<
      HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement
    >
  ) => {
    const { name, value } = e.target;
    setContactData((prevValues) => ({
      ...prevValues,
      [name]: value,
    }));
  };

  return (
    <div className={classes['contact-form__container']}>
      <h1>Contact Form</h1>
      <CustomInput
        name="name"
        type="text"
        placeHolder="Name"
        value={contactData.name}
        onChange={handleInputChange}
      />
      <CustomInput
        name="email"
        type="email"
        placeHolder="Email"
        value={contactData.email}
        onChange={handleInputChange}
      />
      <CustomInput
        name="subject"
        type="text"
        placeHolder="Subject"
        onChange={handleInputChange}
        value={contactData.subject}
      />
      <textarea
        name="message"
        placeholder="Please write here..."
        onChange={handleInputChange}
        value={contactData.message}
      />
      <div className={classes['buttons__wrapper']}>
        <button type="submit">Submit</button>
        <button onClick={handleCancelClick}>Cancel</button>
      </div>

      {/* <div className={classes['logo']}>
        <h1>H</h1>
      </div> */}
    </div>
  );
};

export default ContactPage;
