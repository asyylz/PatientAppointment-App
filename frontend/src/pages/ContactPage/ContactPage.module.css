@import url('./../../App.css');

.contact-form__container {
    display: grid;
    grid-template-columns: 50vw;
    grid-template-rows: 30px repeat(3, 50px) 200px;
    margin: 2rem auto;
    -webkit-animation: scaleUp .7s ease both;
    animation: scaleUp .7s ease both;

}

.contact-form__container h1 {
    font-size: clamp(12px, 2vw, 2.5vw);
}

.buttons__wrapper {
    display: flex;
    margin-bottom: 3rem;
}


@media (min-width: 375px) and (max-width:834px) {
    .contact-form__container {
        grid-template-columns: 80vw;
    }
}

@media (min-width: 375px) and (max-width:413px) {}

.logo {
    height: 100px;
    width: 100px;
    background-color: #8FC1B5;
    border-radius: 50%;
    display: flex;
    justify-content: center;
    align-items: center;
    border: 3px solid #E2D8D7;
    box-shadow: rgba(50, 50, 93, 0.25) 0px 30px 60px -12px inset, rgba(0, 0, 0, 0.3) 0px 18px 36px;
}

.logo h1 {
    font-size: 56px;
    font-family: 'Times New Roman', Times, serif;
}