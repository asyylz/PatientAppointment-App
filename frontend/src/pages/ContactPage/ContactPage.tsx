import React, { useState } from 'react';
import CustomInput from '../../components/CustomInput/CustomInput';
import classes from './ContactPage.module.css';
import useHttp from '../../hooks/useHttp/useHttp';


const ContactPage: React.FC = () => {
  const { submitContactForm } = useHttp();

  const [contactData, setContactData] = useState<ContactFormData>({
    name: '',
    email: '',
    subject: '',
    message: '',
  });

  const handleCancelClick = () => {
    setContactData({
      name: '',
      email: '',
      subject: '',
      message: '',
    });
  };
  const handleInputChange = (
    e: React.ChangeEvent<
      HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement
    >
  ) => {
    const { name, value } = e.target;
    setContactData((prevValues) => ({
      ...prevValues,
      [name]: value,
    }));
  };
  const handleSubmit = async () => {
    const response = await submitContactForm(contactData);
    console.log('asiye');
    if (response.status === 'success') {
      console.log('asiye');
      setContactData({
        name: '',
        email: '',
        subject: '',
        message: '',
      });
    }
  };

  return (
    <div className={classes['contact-form__container']}>
      <h1>Contact Form</h1>
      <CustomInput
        name="name"
        type="text"
        placeHolder="Name"
        value={contactData.name}
        onChange={handleInputChange}
      />
      <CustomInput
        name="email"
        type="email"
        placeHolder="Email"
        value={contactData.email}
        onChange={handleInputChange}
      />
      <CustomInput
        name="subject"
        type="text"
        placeHolder="Subject"
        onChange={handleInputChange}
        value={contactData.subject}
      />
      <textarea
        name="message"
        placeholder="Please write here..."
        onChange={handleInputChange}
        value={contactData.message}
      />
      <div className={classes['buttons__wrapper']}>
        <button type="submit" onClick={handleSubmit}>
          Submit
        </button>
        <button onClick={handleCancelClick}>Cancel</button>
      </div>
    </div>
  );
};

export default ContactPage;
