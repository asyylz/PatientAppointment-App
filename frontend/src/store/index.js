import { configureStore } from '@reduxjs/toolkit';

const store = configureStore({
  reducer: { departments: departmentsSlice.reducer }, // root reducer
});

export default store;
