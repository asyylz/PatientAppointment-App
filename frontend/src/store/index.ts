import { configureStore, combineReducers } from '@reduxjs/toolkit';
import {
  persistReducer,
  persistStore,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import departmentsReducer from './departments-slice';
import doctorsReducer from './doctors-slice';
import reviewsReducer from './reviews-slice';
import currentUserReducer from './currentUser-slice';
//import { useDispatch } from 'react-redux';
import searchReducer from './search-slice';
import appointmentsForDoctorReducer from './appointmentsForDoctor-slice';
import appointmentsForPatientReducer from './appointmentsForPatient-slice';
import storage from 'redux-persist/lib/storage'; // defaults to localStorage for web

// const rootPersistConfig = {
//   key: 'root',
//   storage,
// };

// const persistedReducer = persistReducer(rootPersistConfig, currentUserReducer);

// const store = configureStore({
//   reducer: {
//     currentUser: persistedReducer,
//     departments: departmentsReducer,
//     doctors: doctorsReducer,
//     reviews: reviewsReducer,
//     search: searchReducer,
//     appointmentsForDoctor: appointmentsForDoctorReducer,
//     appointmentsForPatient: appointmentsForPatientReducer,
//   },
//   middleware: (getDefaultMiddleware) =>
//     getDefaultMiddleware({
//       serializableCheck: {
//         ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
//         ignoredActionPaths: ['meta.arg', 'payload.timestamp'],
//         ignoredPaths: ['items.dates'],
//       },
//     }),
//   devTools: process.env.NODE_ENV !== 'production',
// });

// const rootPersistConfig = {
//   key: 'root',
//   storage,
//   whitelist: ['currentUser', 'doctors'], // Only persist the 'currentUserSlice' slice
// };

const authPersistConfig = {
  key: 'currentUser',
  storage: storage,
  // blacklist: ['status', 'error'],
};
const doctorPersistConfig = {
  key: 'doctors',
  storage: storage,
  // blacklist: ['status','error']
};
export const allReducers = combineReducers({
  currentUser: persistReducer(authPersistConfig, currentUserReducer),
  departments: departmentsReducer,
  doctors: persistReducer(doctorPersistConfig, doctorsReducer),
  reviews: reviewsReducer,
  search: searchReducer,
  appointmentsForDoctor: appointmentsForDoctorReducer,
  appointmentsForPatient: appointmentsForPatientReducer,
});

//const persistedReducer = persistReducer(rootPersistConfig, allReducers);

const store = configureStore({
  reducer: allReducers,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
  devTools: process.env.NODE_ENV !== 'production',
});

export type AppDispatch = typeof store.dispatch;
//export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export type AppStore = typeof store;
export type RootState = ReturnType<AppStore['getState']>;

export const persistor = persistStore(store);

export default store;
