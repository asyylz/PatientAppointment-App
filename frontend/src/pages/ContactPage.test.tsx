import renderer from 'react-test-renderer';
import ContactPage from './ContactPage';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import '@testing-library/jest-dom';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import store from './../store/index';
import { submitContactFormMock } from './../_testUtils/mocks/mockHttp';

jest.mock('./../hooks/useHttp');

describe('ContactPage', () => {
  beforeEach(() => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <ContactPage />
        </MemoryRouter>
      </Provider>
    );
  });
  /* -------------------------- - ------------------------- */
  it('1--Matches the DOM snapshot', () => {
    let domTree;
    try {
      domTree = renderer
        .create(
          <Provider store={store}>
            <MemoryRouter>
              <ContactPage />
            </MemoryRouter>
          </Provider>
        )
        .toJSON();
    } catch (error) {
      console.error('Rendering error:', error);
    }

    expect(domTree).toMatchSnapshot();
  });
  /* -------------------------- - ------------------------- */
  it('2--Should render contact form', () => {
    expect(screen.getByPlaceholderText('Name')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Email')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Subject')).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText('Please write here...')
    ).toBeInTheDocument();
    expect(screen.getByRole('button', { name: 'Submit' }));
  });
  /* -------------------------- - ------------------------- */
  it('3--Contact form fields are typed into and form submission is triggered', async () => {
    const nameInput = screen.getByPlaceholderText('Name');
    const emailInput = screen.getByPlaceholderText('Email');
    const subjectInput = screen.getByPlaceholderText('Subject');
    const messageInput = screen.getByPlaceholderText('Please write here...');
    const submitButton = screen.getByRole('button', { name: 'Submit' });

    fireEvent.change(nameInput, { target: { value: 'Alice' } });
    fireEvent.change(emailInput, { target: { value: 'alice@test.com' } });
    fireEvent.change(subjectInput, { target: { value: 'Subject Test' } });
    fireEvent.change(messageInput, { target: { value: 'Test message' } });
    fireEvent.click(submitButton);
    expect(submitContactFormMock).toHaveBeenCalledTimes(1);
    expect(submitContactFormMock).toHaveBeenCalledWith({
      name: 'Alice',
      email: 'alice@test.com',
      subject: 'Subject Test',
      message: 'Test message',
    });

    await waitFor(() => {
      expect(screen.getByPlaceholderText('Name')).toHaveValue('');
      expect(screen.getByPlaceholderText('Email')).toHaveValue('');
      expect(screen.getByPlaceholderText('Subject')).toHaveValue('');
      expect(screen.getByPlaceholderText('Please write here...')).toHaveValue(
        ''
      );
    });
  });
  /* -------------------------- - ------------------------- */
  it('3--Contact form fields are typed but canceled not to submit', async () => {
    const nameInput = screen.getByPlaceholderText('Name');
    const emailInput = screen.getByPlaceholderText('Email');
    const subjectInput = screen.getByPlaceholderText('Subject');
    const messageInput = screen.getByPlaceholderText('Please write here...');
    const cancelButton = screen.getByRole('button', { name: 'Cancel' });

    fireEvent.change(nameInput, { target: { value: 'Alice' } });
    fireEvent.change(emailInput, { target: { value: 'alice@test.com' } });
    fireEvent.change(subjectInput, { target: { value: 'Subject Test' } });
    fireEvent.change(messageInput, { target: { value: 'Test message' } });
    fireEvent.click(cancelButton);

    await waitFor(() => {
      expect(screen.getByPlaceholderText('Name')).toHaveValue('');
      expect(screen.getByPlaceholderText('Email')).toHaveValue('');
      expect(screen.getByPlaceholderText('Subject')).toHaveValue('');
      expect(screen.getByPlaceholderText('Please write here...')).toHaveValue(
        ''
      );
    });
  });
});
