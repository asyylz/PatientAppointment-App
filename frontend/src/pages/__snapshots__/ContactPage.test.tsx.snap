// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactPage 1--Matches the DOM snapshot 1`] = `
<div
  className="contact-form__container"
>
  <h1>
    Contact Form
  </h1>
  <div
    className="wrapper input"
    style={{}}
  >
    <input
      name="name"
      onChange={[Function]}
      placeholder="Name"
      style={{}}
      type="text"
      value=""
    />
  </div>
  <div
    className="wrapper input"
    style={{}}
  >
    <input
      name="email"
      onChange={[Function]}
      placeholder="Email"
      style={{}}
      type="email"
      value=""
    />
  </div>
  <div
    className="wrapper input"
    style={{}}
  >
    <input
      name="subject"
      onChange={[Function]}
      placeholder="Subject"
      style={{}}
      type="text"
      value=""
    />
  </div>
  <textarea
    name="message"
    onChange={[Function]}
    placeholder="Please write here..."
    value=""
  />
  <div
    className="buttons__wrapper"
  >
    <button
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
    <button
      onClick={[Function]}
    >
      Cancel
    </button>
  </div>
</div>
`;
